Sonic SwapX AMO Strategy

naddison36 opened the pull request:
## Changes

* Add new `SonicSwapXAMOStrategy` contract
* Deploy `OSonicHarvester` for collecting SWPx rewards and sending them to the Guardian account
* Upgrade Base Curve AMO Strategy so `withdrawalAll` will work with no assets
* Upgrade Base Vault with complete async withdrawal functionality. eg prevent depositing WETH reserved for the withdrawal queue to a strategy.
* Upgrade Mainnet OETH Vault to support multiple AMOs
* Fixed address in `OETHVaultAdmin`'s `StrategyAddedToMintWhitelist` and `StrategyRemovedFromMintWhitelist` events

![sonicContracts](https://github.com/user-attachments/assets/2eeacb91-6b82-4e07-8e83-abca6a23e7fb)

The inheritance of the Vault contracts

![VaultHierarchy](https://github.com/user-attachments/assets/3b933682-28c8-43d6-aa0d-aff64c68fa6b)

## Testing

Fork tests are in `contracts/test/strategies/sonic/swapx-amo.sonic.fork-test.js`

```zsh
# Start a local forked node of Sonic with
yarn run node:sonic
```

```zsh
# Run the fork tests
yarn run test:sonic-fork
```

## Deployment

As there are changes to the Vault there are three deploy scripts on Sonic, Base and Mainnet

### Sonic

The Sonic deploy script is `contracts/deploy/sonic/009_swapx_amo.js`

```zsh
yarn run deploy:sonic 
```

### Base

The Sonic deploy script is `contracts/deploy/base/027_base_curve_amo_upgrade.js`

```zsh
yarn run deploy:base 
```

### Mainnet

The Sonic deploy script is `contracts/deploy/base/126_oeth_vault_upgrade.js`

```zsh
yarn run deploy:mainnet 
```

## Code Change Checklist

To be completed before internal review begins:

- [ ]  The contract code is complete
- [ ]  Executable deployment file
- [ ]  Fork tests that test after the deployment file runs
- [ ]  Unit tests *if needed
- [ ]  The owner has done a [full checklist review](https://github.com/OriginProtocol/security/blob/master/templates/Contract-Code-Review.md) of the code + tests

Internal review:

- [ ] Two approvals by internal reviewers

sparrowDom reviewed (commented):
I am still reviewing, just posting the first batch of comments

sparrowDom commented on contracts/contracts/strategies/sonic/SonicSwapXAMOStrategy.sol:
```diff
@@ -0,0 +1,707 @@
+     */
+    function _solvencyAssert() internal view {
+        uint256 _totalVaultValue = IVault(vaultAddress).totalValue();
+        uint256 _totalOSSupply = IERC20(os).totalSupply();
```
This will probably never be 0. But if it is we divide by 0

naddison36 commented on contracts/contracts/strategies/sonic/SonicSwapXAMOStrategy.sol:
```diff
@@ -0,0 +1,707 @@
+        } else if (diffBefore < 0) {
+            // If the pool was originally imbalanced in favor of OS, then
+            // we want to check that the pool is now more balanced
+            require(diffAfter <= 0, "Assets overshot peg");
```
Yes, this is the most likely scenario. There is more OS in the pool and `swapAssetsToPool` is used to burn OS from the pool. If we burn too much OS, then the pool can end up with more wS than OS. This check prevents this.

This is the same code used for the Curve AMOs on Mainnet and Base.

naddison36 replied:
Deposits and withdraws are more profitable when the pool is balanced so I think it's ok to bring OS to 1:1 peg and then let the market pull it back to a fair discount given the withdrawal times.

naddison36 commented on contracts/deploy/sonic/009_swapx_amo.js:
```diff
@@ -0,0 +1,145 @@
+
+    const cDripperProxy = await ethers.getContract("OSonicDripperProxy");
+
+    const initSonicStakingStrategy = cHarvester.interface.encodeFunctionData(
```
Good pick. Fixed

naddison36 commented on contracts/contracts/strategies/sonic/SonicSwapXAMOStrategy.sol:
```diff
@@ -0,0 +1,707 @@
+pragma solidity ^0.8.0;
+
+/**
+ * @title SwapX Automated Market Maker (AMO) Strategy
```
I always get that wrong. Thanks. Fixed

naddison36 commented on contracts/contracts/strategies/sonic/SonicSwapXAMOStrategy.sol:
```diff
@@ -0,0 +1,707 @@
+
+        // Safety check that we are dealing with the correct pool tokens
+        require(
+            (_tokenIn == ws || _tokenIn == os) &&
```
Nice suggestion. I've used it.

naddison36 commented on contracts/contracts/strategies/sonic/SonicSwapXAMOStrategy.sol:
```diff
@@ -0,0 +1,707 @@
+    /// @dev Calculate the value of a LP position in a SwapX stable pool
+    /// if the pool was balanced.
+    /// @param lpTokens Amount of LP tokens in the SwapX pool
+    /// @return value The wS value of the LP tokens in the pool
```
That's true. I'll update the Natspec

naddison36 commented on contracts/contracts/strategies/sonic/SonicSwapXAMOStrategy.sol:
```diff
@@ -0,0 +1,707 @@
+        require(_wS == ws, "Unsupported asset");
+
+        // Calculate the required amount of OS to mint based on the wS amount.
+        uint256 osLiquidity = _calcTokensToMint(_wsAmount);
```
Good point. I was intending to be the amount added to the pool but liquidity is misleading for concentrated pools. 
I think `osDepositAmount` is better so will go with that.

naddison36 commented on contracts/contracts/strategies/sonic/SonicSwapXAMOStrategy.sol:
```diff
@@ -0,0 +1,707 @@
+    function _approveBase() internal {
+        // Approve pool for OS and wS (required for adding liquidity)
+        // slither-disable-next-line unused-return
+        IERC20(os).approve(platformAddress, type(uint256).max);
```
Excellent point. I've removed these

naddison36 commented on contracts/contracts/strategies/sonic/SonicSwapXAMOStrategy.sol:
```diff
@@ -0,0 +1,707 @@
+     */
+    modifier skimPool() {
+        IPair(pool).skim(address(this));
+        _;
```
The idea is to prevent donation attacks against the pool from messing with the strategy's calculations. That is, someone front-runs a strategy's operation against the pool so there is more tokens in the pool than the strategy was expecting. The strategy is using the reserve values rather than pool balances.

Most of the time others donating to the pool will end up benefiting the strategy but there may be some cases where it throws out the strategy. At the very least, they can donate to force reverts.

There should only be dust amounts left over from deposit / withdraw but we could also skim those. Gas is cheap on Sonic.

sparrowDom replied:
Skim returns the token0 & token1 to the caller, and not the LP token. Meaning that this applies to `deposit`, `depositAll` and the 2 swap transactions. Which one of these do you think an attacker could manipulate by donating tokens to?

DanielVF replied:
<img width="429" alt="image" src="https://github.com/user-attachments/assets/dfa335eb-9136-4721-b83b-829929b2ef79" />

If we skim before or after, we don't lose our funds depositing. The deposit works off reserve ratios, we work off reserves.

If we skim before or after, we get any extra funds on the contract that someone else left.

However, only if we skim after do we get any of our own funds back from us not calculating correctly on a deposit. So it seems like we should skim after.

naddison36 commented on contracts/contracts/strategies/sonic/SonicSwapXAMOStrategy.sol:
```diff
@@ -0,0 +1,707 @@
+        // Emit event for the deposited wS tokens
+        emit Deposit(_wS, pool, _wsAmount);
+        // Emit event for the minted OS tokens
+        emit Deposit(os, pool, osLiquidity);
```
We have emitted `Deposit` and `Withdrawal` events for the OTokens in the Mainnet and Base Curve AMOs. It helps the analytics track the OTokens being minted and burned.

sparrowDom replied:
I see, my mistake... thanks

naddison36 commented on contracts/contracts/strategies/sonic/SonicSwapXAMOStrategy.sol:
```diff
@@ -0,0 +1,707 @@
+        // Emit event for the withdrawn wS tokens
+        emit Withdrawal(_asset, pool, _wsAmount);
+        // Emit event for the burnt OS tokens
+        emit Withdrawal(os, pool, osToBurn);
```
This is the approach taken for other AMOs

naddison36 commented on contracts/contracts/strategies/sonic/SonicSwapXAMOStrategy.sol:
```diff
@@ -0,0 +1,707 @@
+        // Emit event for the withdrawn wS tokens
+        emit Withdrawal(ws, pool, wsBalance);
+        // Emit event for the burnt OS tokens
+        emit Withdrawal(os, pool, osToBurn);
```
The same is done in the rebalancing functions of the OETH Curve AMO. eg `mintAndAddOTokens` emits a `Deposit` event for minted OETH and emits a `Withdrawal` event in `removeAndBurnOTokens` for burnt OETH.

naddison36 commented on contracts/contracts/strategies/sonic/SonicSwapXAMOStrategy.sol:
```diff
@@ -0,0 +1,707 @@
+        // Perform the swap on the pool
+        IPair(pool).swap(amount0, amount1, address(this), new bytes(0));
+
+        // The slippage protection against the amount out is indirectly done
```
The design assumes the same as the Curve AMOs in it rebalancing should always move closer to balanced.

naddison36 commented:
Developer internal security checklist

## Requirements

This PR adds an AMO strategy to the Origin Vault on Sonic (OS) that integrates to SwapX's stable pool.

It also deploys a Harvester contract to collect SWPx rewards and transfer them to the Guardian multi-sig.

## Easy Checks

#### Authentication

- [x] Never use tx.origin
- [x] Every external/public function is supposed to be externally accessible
- [x] Every external/public function has the correct authentication

#### Ethereum

- [x] Contract does not send or receive Ethereum.
- [x] Contract has no payable methods.
- [x] Contract is not vulnerable to being sent self destruct ETH

#### Cryptographic code

- [x] This contract code does not roll it's own crypto.
- [x] No signature checks without reverting on a 0x00 result.
- [x] No signed data could be used in a replay attack, on our contract or others.

#### Gas problems

- [x] Contracts with for loops must have either:
    - [ ] A way to remove items
    - [x] Can be upgraded to get unstuck
    - [ ] Size can only controlled by admins
- [x] Contracts with for loops must not allow end users to add unlimited items to a loop that is used by others or admins.

#### Black magic

- [x] Does not contain `selfdestruct`
- [x] Does not use `delegatecall` outside of proxying. _If an implementation contract were to call delegatecall under attacker control, it could call selfdestruct the implementation contract, leading to calls through the proxy silently succeeding, even though they were failing._
- [x] Address.isContract should be treated as if could return anything at any time, because that's reality.

#### Overflow

- [x] Code is solidity version >= 0.8.0
- [x] All for loops use uint256

#### Proxy

- [x] No storage variable initialized at definition when contract used as a proxy implementation.

#### Events
- [x] All state changing functions emit events

## Medium Checks

#### Rounding and casts
- [ ] Contract rounds in the protocols favor
- [ ] Contract does not have bugs from loosing rounding precision
- [ ] Code correctly multiplies before division
- [ ] Contract does not have bugs from zero or near zero amounts
- [ ] Safecast is aways used when casting

#### Dependencies

- [x] Review any new contract dependencies thoroughly (e.g. OpenZeppelin imports) when new dependencies are added or version of dependencies changes.
- [x] If OpenZeppelin ACL roles are use review & enumerate all of them.
- [x] Check OpenZeppelin [security vulnerabilities](https://github.com/OpenZeppelin/openzeppelin-contracts/security/advisories) and see if any apply to current PR considering the version of OpenZeppelin contract used.

#### External calls

- [x] Contract addresses passed in are validated
- [ ] No unsafe external calls
- [x] Reentrancy guards on all state changing functions
    - [x] Still doesn't protect against external contracts changing the state of the world if they are called.
- [x] No malicious behaviors
- [x] Low level call() must require success.
- [ ] No slippage attacks (we need to validate expected tokens received)
- [x] Oracles, one of:
  - [x] No oracles
  - [ ] Oracles can't be bent
  - [ ] If oracle can be bent, it won't hurt us.
- [x] Do not call balanceOf for external contracts to determine what they will do when they use internal accounting
  - `getReserves` is used on the pool rather than getting the balance of the pool

#### Tests

  - [x] Each publicly callable method has a test
  - [ ] Each logical branch has a test
  - [ ] Each require() has a test
  - [ ] Edge conditions are tested
  - [ ] If tests interact with AMM make sure enough edge cases (pool tilts) are tested. Ideally with fuzzing.

#### Deploy

- [x] Deployer permissions are removed after deploy

## Strategy Specific

_Remove this section if the code being reviewed is not a strategy._

#### Strategy checks

- [x] Check balance cannot be manipulated up AND down by an attacker
- [ ] No read only reentrancy on downstream protocols during checkBalance
- [x] All reward tokens are collected
- [x] The harvester can sell all reward tokens
  - The Harvester transfers SWPx rewards to the Guardian 
- [ ] No funds are left in the contract that should not be as a result of depositing or withdrawing
  - There can be a dust amount of wS left in the contract
- [x] All funds can be recovered from the strategy by some combination of depositAll, withdraw, or withdrawAll()
- [ ] WithdrawAll() can always withdraw an amount equal to or larger than checkBalances report, even in spite of attacker manipulation.
- [ ] WithdrawAll() cannot be MEV'd
- [x] Strategist cannot steal funds

#### Downstream

- [ ] We have monitoring on all backend protocol's governances
- [ ] We have monitoring on a pauses in all downstream systems

## Thinking

#### Logic

  _Are there bugs in the logic?_

  - [ ] Correct usage of global & local variables. -> they might differentiate only by an underscore that can be overlooked (e.g. address vs _address).

#### Deployment Considerations

_Are there things that must be done on deploy, or in the wider ecosystem for this code to work. Are they done?_

#### Internal State

- What can be always said about relationships between stored state
- What must hold true about state before a function can run correctly (preconditions)
- What must hold true about the return or any changes to state after a function has run.

Does this code do that?

#### Attack

_What could the impacts of code failure in this code be._

_What conditions could cause this code to fail if they were not true._

_Does this code successfully block all attacks._

#### Flavor

_Could this code be simpler?_

_Could this code be less vulnerable to other code behaving weirdly?_

sparrowDom commented on contracts/contracts/strategies/sonic/SonicSwapXAMOStrategy.sol:
```diff
@@ -0,0 +1,710 @@
+    /**
+     * @notice Withdraw wS and OS from the SwapX pool, burn the OS,
+     * and transfer the wS to the recipient.
+     * @param _recipient Address to receive withdrawn asset which is normally the Vault.
```
nit: the vault should be the only valid recipient

naddison36 replied:
I've updated the Natspec

sparrowDom commented on contracts/test/strategies/sonic/swapx-amo.sonic.fork-test.js:
```diff
@@ -0,0 +1,1234 @@
+    }
+  };
+
+  async function assertDeposit(wsDepositAmount) {
```
many of these asserts could also test that after the operation the LP token balance of the strategy contract equals 0

sparrowDom replied:
and they could also assert that the strategy has a 0 OS balance after any operation

naddison36 replied:
That's a good suggestion. I've added this to the deposit, withdraw and swap assert functions.

sparrowDom commented on contracts/contracts/strategies/sonic/SonicSwapXAMOStrategy.sol:
```diff
@@ -0,0 +1,710 @@
+    ) external override onlyVault nonReentrant skimPool {
+        require(_wsAmount > 0, "Must withdraw something");
+        require(_asset == ws, "Unsupported asset");
+
```
We should add a require here: 
```
require(_recipient == vaultAddress, "Only withdraw to vault allowed");
```

naddison36 replied:
VaultCore's `redeem` code does allow withdrawing directly to the redeemer. But that's not relevant for Sonic as `redeem` has been disabled.

```Solidity
// Nothing in Vault, but something in Strategy, send from there
IStrategy strategy = IStrategy(strategyAddr);
strategy.withdraw(msg.sender, assetAddr, outputs[i]);
```

I'm inclined to keep the behaviour between the strategies consistent and allow the Vault to withdraw to a recipient.
I can be convinced the other way, though.

sparrowDom replied:
I guess it comes down to, can/will this ever be an asset default strategy. With AerodromAMO we knew that it won't be, so we added [the require check](https://github.com/OriginProtocol/origin-dollar/blob/master/contracts/contracts/strategies/aerodrome/AerodromeAMOStrategy.sol#L889). 

I think any strategy operating on a pool will probably never be an asset default one, and for that reason we could add a require here. Again since only vault calls this code it is a defensive programming measure that is of low importance. The require would basically only prevent the governance mistakingly setting the SwapxAMO as a default strategy. I will leave the decision up to you.

naddison36 replied:
I'm happy to do the same as the Aerodrome AMO. I like being consistent :-)

Given deposits and withdrawals can be front-run, we definitely don't want the AMO strategy set as the default strategy in the vault.

I'll add the require

sparrowDom commented on contracts/contracts/strategies/sonic/SonicSwapXAMOStrategy.sol:
```diff
@@ -0,0 +1,710 @@
+        uint256 _wsAmount
+    ) external override onlyVault nonReentrant skimPool {
+        require(_wsAmount > 0, "Must withdraw something");
+        require(_asset == ws, "Unsupported asset");
```
These 2 requires could have unit tests

naddison36 replied:
Fixed

sparrowDom commented on contracts/contracts/strategies/sonic/SonicSwapXAMOStrategy.sol:
```diff
@@ -0,0 +1,710 @@
+        (uint256 wsReserves, , ) = IPair(pool).getReserves();
+        require(wsReserves > 0, "Empty pool");
+
+        lpTokens = (_wsAmount * IPair(pool).totalSupply()) / wsReserves;
```
To express the rounding up, and not round up you could also do: 
```
lpTokens = (_wsAmount * IPair(pool).totalSupply() + 1e18 - 1) / wsReserves;
```
Pro: doesn't round up when the division is perfect  / whole numbered 
Con: affects code readability negatively

naddison36 replied:
we want to round up to ensure we get enough LP tokens out of the gauge and then wS tokens out off the pool.

I originally had this code but testing showed cases where not enough LP tokens were used to get the required amount of wS tokens.
```
lpTokens =
            ((_wsAmount + 1) * IPair(pool).totalSupply()) /
            ws.balanceOf(pool);
```

sparrowDom commented on contracts/contracts/strategies/sonic/SonicSwapXAMOStrategy.sol:
```diff
@@ -0,0 +1,710 @@
+        returns (uint256 lpTokens)
+    {
+        /* The rate between coins in the pool determines the rate at which pool returns
+         * tokens when doing balanced removal (remove_liquidity call). And by knowing how much wS
```
nit: `burn` is the call instead of `remove_liquidity`

naddison36 replied:
This was copied from the Curve AMO. I've corrected. Thanks

sparrowDom commented on contracts/contracts/strategies/sonic/SonicSwapXAMOStrategy.sol:
```diff
@@ -0,0 +1,710 @@
+    {
+        /* The rate between coins in the pool determines the rate at which pool returns
+         * tokens when doing balanced removal (remove_liquidity call). And by knowing how much wS
+         * we want we can determine how much of OS we receive by removing liquidity.
```
not sure should this comment say: 
"... we want we can determine how much LP we need to burn in order to receive the desired amount of wS
""

naddison36 replied:
This was copied from the Curve AMO which is a little different in Curve supports unbalanced removals.

I've changed it to this
```
  /* The SwapX pool proportionally returns the reserve tokens when removing liquidity.
   * First, calculate the proportion of required wS tokens against the pools wS reserves.
   * That same proportion is used to calculate the required amount of pool LP tokens.
   * For example, if the required wS tokens is 10% of the pool's wS reserves,
   * then 10% of the pool's LP supply needs to be burned.
```

sparrowDom commented on contracts/contracts/strategies/sonic/SonicSwapXAMOStrategy.sol:
```diff
@@ -0,0 +1,710 @@
+        _solvencyAssert();
+
+        // Emit event for the withdrawn wS tokens
+        emit Withdrawal(_asset, pool, _wsAmount);
```
super nit: could be `_ws` instead of `_asset` for tiny code readability improvement

naddison36 replied:
I agree. Fixed

sparrowDom commented on contracts/contracts/strategies/sonic/SonicSwapXAMOStrategy.sol:
```diff
@@ -0,0 +1,710 @@
+     * The OS/wS price is < 1.0 so OS is being bought at a discount.
+     * @param _wsAmount Amount of Wrapped S (wS) to swap into the pool.
+     */
+    function swapAssetsToPool(uint256 _wsAmount)
```
the name could also be `swapWsToPool`

naddison36 replied:
I tried to keep the function name generic in case we reused this strategy for other assets or chains. 
The same was done for the Curve AMO which has `mintAndAddOTokens`, `removeAndBurnOTokens` and `removeOnlyAssets` rebalance functions. This worked out when the Sonic Curve AMO was done.

sparrowDom replied:
makes sense I agree the same function names across multiple strategies is better.

sparrowDom commented on contracts/contracts/strategies/sonic/SonicSwapXAMOStrategy.sol:
```diff
@@ -0,0 +1,710 @@
+        // 3. Add wS and OS back to the pool in proportion to the pool's reserves
+
+        // The wS is from the swap and any wS that was sitting in the strategy
+        uint256 wsDepositAmount = IERC20(ws).balanceOf(address(this));
```
The code below could be replaced by calling `_deposit(wsDepositAmount)` though a new parameter would need to be added so the `_deposit()` doesn't emit events in this case

naddison36 replied:
Good idea. I had to move the `Deposit` events out of the internal `_deposit` function so it makes the external `deposit` and `depositAll` a little messier. But I think it's worth it as it moves an extra four lines of code into an exiting common function.

sparrowDom reviewed (commented):
Another round of comments for the PR. Overall the code is in a really good condition 💪

naddison36 commented on contracts/contracts/strategies/sonic/SonicSwapXAMOStrategy.sol:
```diff
@@ -0,0 +1,710 @@
+     * Initializer for setting up strategy internal state. This overrides the
+     * InitializableAbstractStrategy initializer as SwapX strategies don't fit
+     * well within that abstraction.
+     * @param _rewardTokenAddresses Address of SWPx token
```
Fixed. Thanks

naddison36 commented on contracts/contracts/strategies/sonic/SonicSwapXAMOStrategy.sol:
```diff
@@ -0,0 +1,710 @@
+    }
+
+    function _deposit(address _wS, uint256 _wsAmount) internal {
+        require(_wsAmount > 0, "Must deposit something");
```
Good pickup. I've added a test

naddison36 commented on contracts/contracts/strategies/sonic/SonicSwapXAMOStrategy.sol:
```diff
@@ -0,0 +1,710 @@
+
+    function _deposit(address _wS, uint256 _wsAmount) internal {
+        require(_wsAmount > 0, "Must deposit something");
+        require(_wS == ws, "Unsupported asset");
```
Good pickup. I've added a test

naddison36 commented on contracts/test/strategies/sonic/swapx-amo.sonic.fork-test.js:
```diff
@@ -0,0 +1,1234 @@
+      const { oSonicVaultSigner, swapXAMOStrategy } = fixture;
+
+      // Vault withdraw alls from the strategy
+      const tx = swapXAMOStrategy.connect(oSonicVaultSigner).withdrawAll();
```
There's a `not` in the assertion 
`await expect(tx).to.not.revertedWith("Protocol insolvent");`

sparrowDom replied:
Ah my mistake I'm sorry

naddison36 commented on contracts/contracts/vault/OETHVaultAdmin.sol:
```diff
@@ -19,6 +19,48 @@ contract OETHVaultAdmin is VaultAdmin {
+     * @notice Removes a strategy from the mint whitelist.
+     * @param strategyAddr Strategy address
+     */
+    function removeStrategyFromMintWhitelist(address strategyAddr)
```
I've added a unit tests

naddison36 commented on contracts/contracts/strategies/sonic/SonicSwapXAMOStrategy.sol:
```diff
@@ -0,0 +1,710 @@
+    ****************************************/
+
+    /**
+     * @notice Withdraw wS and OS from the SwapX pool, burn the OS,
```
I've updated the deposit Natspec

naddison36 commented on contracts/deploy/base/027_base_curve_amo_upgrade.js:
```diff
@@ -0,0 +1,69 @@
+    );
+
+    // Deploy Base Curve AMO implementation
+    const dOETHBaseCurveAMOImpl = await deployWithConfirmation(
```
This has already been deployed

sparrowDom commented on contracts/contracts/strategies/sonic/SonicSwapXAMOStrategy.sol:
```diff
@@ -0,0 +1,738 @@
+        (uint256 osDepositAmount, uint256 lpTokens) = _deposit(wsDepositAmount);
+
+        // Emit event for the minted OS tokens
+        emit Deposit(os, pool, osToMint + osDepositAmount);
```
I am not sure adding these 2 amounts together are correct. The `osToMint` was part of the swap transaction and there to swap OS for wS. Once we have the wS the corresponding OS part is minted and deposited. 

Hmm I think I see the thought process, so instead of emitting 2 deposit events 1 for wS amount and 1 for OS amount, we emit only 1 with OS amounts combined.

sparrowDom replied:
I guess this way of doing it is better, since even though we deposit ws&OS into the pool, no WS assets have really moved from the Vault. Only OS has been minted, swapped and deposited. I think the way you did it is more correct.

Maybe a comment to explain the thought process?

naddison36 replied:
The thought process was to not emit three Deposit events. Instead, just emit one Deposit event for wS and the other for OS. But as we mint OS twice, the amounts are added together

naddison36 commented on contracts/deploy/sonic/009_swapx_amo.js:
```diff
@@ -0,0 +1,145 @@
+    );
+
+    // Deploy Sonic SwapX AMO Strategy implementation
+    const dSonicSwapXAMOStrategy = await deployWithConfirmation(
```
I've added a `deploySonicSwapXAMOStrategyImplementation` deploy action

sparrowDom commented:
I've added [a commit](https://github.com/OriginProtocol/origin-dollar/pull/2439/commits/2d62633bca6c964559c3a9b892fdcbdaf506cd18) where a brownie script dpeloys ~5m tokens into the AMO pool and then lets the potential attacker tilt the pool into both directions: OS heavy and WS heavy. `checkBalanace(ws)` is called on the strategy before the tilt and after the tilt. The differences in values are displayed in the below chart: 
<img width="619" alt="Screenshot 2025-03-14 at 10 30 04" src="https://github.com/user-attachments/assets/5f4ddaae-8d19-4f07-b199-11b602fe26fb" />

The amount deposited in the strategy (and returned by `checkBalance`) is roughly `5 * 1e24`. And the differences before and after the tilt are sub 10 WEI.

Which proves that `checkBalance` is resistant to pool manipulation ✅

sparrowDom commented:
## Requirements

This PR creates an AMO strategy for the OS/WS SwapX pool and deploys the harvester to harvest the rewards. PR also upgrades the Base Curve AMO to enable `withdrawAll()` when strategy is empty.

## Easy Checks

#### Authentication

- [X] Never use tx.origin
- [x] Every external/public function is supposed to be externally accessible
- [x] Every external/public function has the correct authentication

#### Ethereum

- [x] Contract does not send or receive Ethereum.
- [x] Contract has no payable methods.
- [x] Contract is not vulnerable to being sent self destruct ETH

#### Cryptographic code

- [x] This contract code does not roll it's own crypto.
- [x] No signature checks without reverting on a 0x00 result.
- [x] No signed data could be used in a replay attack, on our contract or others.

#### Gas problems

_no loops_

#### Black magic

- [x] Does not contain `selfdestruct`
- [x] Does not use `delegatecall` outside of proxying. _If an implementation contract were to call delegatecall under attacker control, it could call selfdestruct the implementation contract, leading to calls through the proxy silently succeeding, even though they were failing._
- [x] Address.isContract should be treated as if could return anything at any time, because that's reality.

#### Overflow

- [x] Code is solidity version >= 0.8.0

#### Proxy

- [x] No storage variable initialized at definition when contract used as a proxy implementation.

#### Events
- [x] All state changing functions emit events

## Medium Checks

#### Rounding and casts
- [x] Contract rounds in the protocols favor
- 🟡  Contract does not have bugs from loosing rounding precision
  - it might have errors especially if we were to operate with tiny numbers. Normal users won't ever interact directly with this contract so it doesn't need to be fuzz tested with various edge case amounts. 
- [x] Code correctly multiplies before division
  - doesn't in `_invariant` function, though the code is copied from the SwapX pool, and it is more important that it can mimic the results SwapX pool produces
- 🟡 Contract does not have bugs from zero or near zero amounts
  - it might have, though we won't test for it, since this won't be an asset default (user interact-able) strategy
- [x] Safecast is aways used when casting

#### Dependencies
- [x] Review any new contract dependencies thoroughly (e.g. OpenZeppelin imports) when new dependencies are added or version of dependencies changes.
- [x] Check OpenZeppelin [security vulnerabilities](https://github.com/OpenZeppelin/openzeppelin-contracts/security/advisories) and see if any apply to current PR considering the version of OpenZeppelin contract used.

#### External calls

- [x] Contract addresses passed in are validated
- [x] No unsafe external calls
- [x] Reentrancy guards on all state changing functions
    - [ ] Still doesn't protect against external contracts changing the state of the world if they are called.
- [x] No malicious behaviors
- [ ] ~~Low level call() must require success.~~ no such calls
- [x] No slippage attacks (we need to validate expected tokens received)
  - we deposit / withdraw in a balanced manner and using Vault value checker prevents any slippage attacks
  - when swapping only the changes towards more balanced pool are allowed, which prevents slippage attacks
- [x] Oracles, one of:
  - [x] No oracles
  - [ ] Oracles can't be bent
  - [ ] If oracle can be bent, it won't hurt us.
- 🟡  Do not call balanceOf for external contracts to determine what they will do when they use internal accounting
  - we do need to call `balanceOf` on tokens and the gauge

#### Tests

  - [x] Each publicly callable method has a test
  - 🟣  Each require() has a test & Each logical branch has a test
    - "No LP tokens to burn" in `swapAssetsToPool` Maybe if `_wsAmount` is 1 WEI this will get triggered
    - "Too much OS in strategy" in `swapOTokensToPool`. Could be triggered in a test by donating OS to the strategy
    - "Unsupported asset" in `checkBalance`
    - "Position balance is worsened" in `improvePoolBalance` (though this one might be tricky to setup - lets skip it)
    - "Not enough wS removed from pool" in `withdraw` (I would skip this one as well)
    - "Empty pool" `_calcTokensToMint` (I'd skip this test)
    - "Unsupported swap" in `_swapExactTokensForTokens`. Though I think that code is unreachable
  - [x] Edge conditions are tested
  - 🟡  If tests interact with AMM make sure enough edge cases (pool tilts) are tested. Ideally with fuzzing.
    - we don't have fuzzing, but there are some fork tests tilting the pool.

#### Deploy

- [x] Deployer permissions are removed after deploy

## Strategy Specific
#### Strategy checks

- [x] Check balance cannot be manipulated up AND down by an attacker
   - this has been confirmed with brownie simulations posted one message above
- [x] No read only reentrancy on downstream protocols during checkBalance
- [x] All reward tokens are collected
- [x] The harvester can sell all reward tokens
- [x] No funds are left in the contract that should not be as a result of depositing or withdrawing
  - there can be dust amounts of WS due to rounding up when removing liquidity
- [x] All funds can be recovered from the strategy by some combination of depositAll, withdraw, or withdrawAll()
- [x] WithdrawAll() can always withdraw an amount equal to or larger than checkBalances report, even in spite of attacker manipulation.
- [x] WithdrawAll() cannot be MEV'd
  - not when using Vault Value checker
- [x] Strategist cannot steal funds

#### Downstream

- [ ] We have monitoring on all backend protocol's governances
- [ ] We have monitoring on a pauses in all downstream systems

## Thinking

#### Logic

Logic looks ok. 

  - [x] Correct usage of global & local variables. -> they might differentiate only by an underscore that can be overlooked (e.g. address vs _address).

#### Deployment Considerations

Pool and gauge are already setup. Nothing further needs to happen

#### Internal State

The strategy holds various configuration parameters in its storage state. There aren't any states that change within the strategy contract during the operation that would be important.

#### Attack

`Deposit`, `withdrawal` and `withdrawAll` can all be MEV-ed. That is why it is important that we maintain the use of VaultValueChecked when depositing / withdrawing to/from the strategy. 

#### Flavor

Code is really clean.

sparrowDom commented:
I've altered the simulation to more drastically tilt the pool when comparing the `checkBalance` values. [Here is](https://gist.github.com/sparrowDom/f5f0a6b619c447f28c345e8e548968da) the raw data which includes values like pool token balances before and after the pool tilt. The differences with `checkBalance` values become more drastic with bigger tilts: 
![checkBalance](https://github.com/user-attachments/assets/3d7892ee-cebb-4cbe-8ef0-f42bbd664464)

DanielVF commented on contracts/contracts/strategies/sonic/SonicSwapXAMOStrategy.sol:
```diff
@@ -0,0 +1,738 @@
+        // Emit event for the deposited wS tokens
+        emit Deposit(ws, pool, _wsAmount);
+        // Emit event for the minted OS tokens
+        emit Deposit(os, pool, osDepositAmount);
```
The deposit and withdraw events are from the vaults perspective. What user funds is the vault investing and withdrawing. We don't emit events in our AMO's for minted protocol owned liquidity.

(Applies to all deposit/withdraw events in this contract)

naddison36 replied:
The Curve AMO emits Deposit and Withdraw events for protocol owned liquidity.

Here's the `Deposit` event for OETH in the OETH Curve AMO https://github.com/OriginProtocol/origin-dollar/blob/b4544dce7751298067b18084ac326142d82aef13/contracts/contracts/strategies/ConvexEthMetaStrategy.sol#L212

The `Deposit` event is also emitted in the `mintAndAddOTokens` rebalance function https://github.com/OriginProtocol/origin-dollar/blob/b4544dce7751298067b18084ac326142d82aef13/contracts/contracts/strategies/ConvexEthMetaStrategy.sol#L414

DanielVF commented on contracts/contracts/strategies/sonic/SonicSwapXAMOStrategy.sol:
```diff
@@ -0,0 +1,738 @@
+        (uint256 osDepositAmount, uint256 lpTokens) = _deposit(wsDepositAmount);
+
+        // Emit event for the minted OS tokens
+        emit Deposit(os, pool, osToMint + osDepositAmount);
```
In our current contract code and analytics, swaps and rebalances don't emit events.

naddison36 replied:
The Curve AMO does emit Deposit/Withdraw events for the OToken

DanielVF commented on contracts/contracts/strategies/sonic/SonicSwapXAMOStrategy.sol:
```diff
@@ -0,0 +1,738 @@
+
+        // 3. Add wS and OS back to the pool in proportion to the pool's reserves
+        (uint256 osDepositAmount, uint256 lpTokens) = _deposit(wsDepositAmount);
+
```
Missing solvency check here.

naddison36 replied:
The solvency check is in the internal `_deposit` function

naddison36 replied:
I've moved `_solvencyAssert` out of the internal `_deposit` to the external `deposit`, `depositAll` and `swapOTokensToPool` functions so it's clearer.

DanielVF commented on contracts/contracts/strategies/sonic/SonicSwapXAMOStrategy.sol:
```diff
@@ -0,0 +1,738 @@
+        require(wsReserves > 0, "Empty pool");
+
+        // OS to add = (wS being added * OS in pool) / wS in pool
+        osAmount = (_wsAmount * osReserves) / wsReserves;
```
🟥 🟥 rekt, rekt.

Attacker front runs a deposit transaction by depositing a huge amount on either side. We deposit matching. When the pool is imbalanced, you get less LP per funds on deposit. (You get more on a withdraw). We lose money, get back less LP tokens. Attacker rebalances pool, has more LP tokens (since they profited from us getting less).

naddison36 replied:
Yes, deposits can be front-run which is why they need to be wrapped with the vault value checker.

naddison36 replied:
I've created PR https://github.com/OriginProtocol/origin-dollar/pull/2464 to add front-running protect on deposits

DanielVF commented:
@sparrowDom How can an attacker MEV attack withdraw and WithdrawAll()?

naddison36 commented on contracts/contracts/strategies/sonic/SonicSwapXAMOStrategy.sol:
```diff
@@ -0,0 +1,738 @@
+        override
+        onlyVault
+        nonReentrant
+        skimPool
```
I thought it was confusing having modifiers at different levels which is why I kept them all at the external function level.

`skimPool` can be moved from `deposit` and `depositAll` to `_deposit`.

`skimPool` needs to remain at the `withdraw` level as the pool reserves are read before the internal `_withdrawFromGaugeAndPool` is called.
`withdrawAll` doesn't read the reserves before `_withdrawFromGaugeAndPool`.
`swapAssetsToPool` need the `skimPool` modifier as the reserves are read before `_swapExactTokensForTokens` is called.
`swapOTokensToPool` is different in it doesn't read reserves before `_swapExactTokensForTokens`.

naddison36 commented:
> @sparrowDom How can an attacker MEV attack withdraw and WithdrawAll()?

@DanielVF I've added more fork tests and shown `withdraw` and `withdrawAll` can not be front-run attacked like deposit. The strategy profits if the pool is tilted away from being balanced in either direction before a withdraw.

I've updated the `withdraw` and `withdrawAll` Natspec

sparrowDom commented:
@DanielVF yes you are right, my mistake. Any bends of the pool before a withdrawal should only yield profit

sparrowDom commented:
I've done the deposit simulation tests. There are slight -1 to -3 WEI losses when depositing, depending on the pool tilt. Since the pool allows deposits only in the 0.5% deviation from the balanced state the more tilted deposits are reverted indicated by the 0 deposit profits in the graph. 
![deposit](https://github.com/user-attachments/assets/4c50d3ab-fb18-4ddf-9e1d-efb5313b28d8)
